### PR TITLE
Rule 2.5: Consider macros accessed before definition

### PR DESCRIPTION
**Pull request by lcartey**
## Description

Fixes https://github.com/github/codeql-coding-standards/issues/776.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - `RULE-2-5`

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: c/misra/src/rules/RULE-2-5/UnusedMacroDeclaration.ql [hunk 1]
```diff
@@ -19,7 +19,18 @@ import codingstandards.c.misra
...
   not exists(MacroAccess ma | ma.getMacro() = m) and
+  // Or if there exists a check whether the macro is defined which the extractor
+  // hasn't been able to tie to a macro (usually because this use came before
+  // the macro was defined e.g. a header guard)
+  not exists(PreprocessorBranchDirective bd |
+    // Covers the #ifdef and #ifndef cases
+    bd.getHead() = m.getName()
+    or
+    // Covers the use of defined() to check if a macro is defined
+    m.getName() = bd.getHead().regexpCapture(".*defined\\((.*)\\).*", 1)
```
Comment (nicolaswill) [inline c/misra/src/rules/RULE-2-5/UnusedMacroDeclaration.ql:32]: I think it's worth making a small change here to better handle whitespace and missing parenthesis.

Some example test-cases, which should all have equivalent functionality:
`#if defined(VAR)`

`#if defined ( VAR )`

`#if defined VAR`

Comment (nicolaswill) [inline c/misra/src/rules/RULE-2-5/UnusedMacroDeclaration.ql:32]: Correction to my previous (now-deleted) suggestion, as only spaces (not newlines) are valid syntax here:
```suggestion
    m.getName() = bd.getHead().regexpCapture(".*defined *\\(? *([^\\s()]+) *\\)?\\.*", 1)
```

**Review by nicolaswill — CHANGES_REQUESTED** [review]

File: c/misra/src/rules/RULE-2-5/UnusedMacroDeclaration.ql (hunk 1 above)
Comment (lcartey) [reply c/misra/src/rules/RULE-2-5/UnusedMacroDeclaration.ql]: Agreed, changed 👍

Comment (lcartey) [reply c/misra/src/rules/RULE-2-5/UnusedMacroDeclaration.ql]: Done - this required disabling clang-format for this section.

**Review by nicolaswill — APPROVED** [review]

**Pull request merged**